### PR TITLE
Dynamically change color scheme from color_scheme query param

### PR DESCRIPTION
**Pull request by moon0326**
Related to https://github.com/woocommerce/woocommerce/issues/52875

## Proposed Changes

* Support changing the color scheme based on the `color_scheme` query parameter when the user is using the new passwordless Jetpack connection

## Testing Instructions

1. Checkout this branch locally and run `yarn start`
2. Once the build is ready, visit this [link](http://calypso.localhost:3000/jetpack/connect/authorize?client_id=239172539&redirect_uri=https%3A%2F%2Fpink-piccolo.jurassic.ninja%2Fwp-admin%2Fadmin.php%3Fhandler%3Djetpack-connection-webhooks%26action%3Dauthorize%26_wpnonce%3Df3099546f0%26redirect%3Dhttps%253A%252F%252Fpink-piccolo.jurassic.ninja%252Fwp-admin%252Fadmin.php%253Fpage%253Dwc-admin&state=1&scope=administrator%3A1a2e2b51282f8ac685ca15e67964f8f4&user_email=moon.kyong%40automattic.com&user_login=moon0326&jp_version=14.1-a.3&secret=QYBTIYTgqLIIMcpg68zLcFnH8pyT9Jmf&blogname=Pink%20Piccolo&site_url=https%3A%2F%2Fpink-piccolo.jurassic.ninja&home_url=https%3A%2F%2Fpink-piccolo.jurassic.ninja&site_icon=&site_lang=en_US&site_created=2024-11-19%2018%3A10%3A55&allow_site_connection=1&calypso_env=production&source=&_as=wp-cli&locale=en&from=woocommerce-core-profiler&plugin_name=woocommerce-payments&color_scheme=sunrise&_ui=189042142&_ut=wpcom%3Auser_id&purchase_nonce=3UiGhvFH&_wp_nonce=d1d5864992&redirect_after_auth=https%3A%2F%2Fpink-piccolo.jurassic.ninja%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin&site=https%3A%2F%2Fpink-piccolo.jurassic.ninja)
3. Confirm the color scheme is `sunrise`
4. Copy the link and search for `color_scheme`
5. Change it to one of the supported color schemes and revisit the link.
6. Confirm the color scheme change.

Supported color schemes:

- default
- light
- modern
- blue
- coffee
- ectoplasm
- midnight
- ocean
- sunrise

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/layout/color-scheme/index.jsx [hunk 1]
```diff
@@ -16,6 +16,12 @@ export function getColorScheme( { state, isGlobalSidebarVisible, sectionName } )
 	return siteColorScheme ?? calypsoColorScheme;
 }
 
+export function getColorSchemeFromCurrentQuery( currentQuery ) {
+	return currentQuery?.color_scheme
+		? currentQuery?.color_scheme
+		: new URLSearchParams( currentQuery?.redirect_to ).get( 'color_scheme' );
```
Comment (chihsuan) [inline client/layout/color-scheme/index.jsx]: If redirect_to is null, undefined, or an invalid URL, URLSearchParams will throw an error. Should we add a fallback for cases where redirect_to is not a valid URL?

File: client/layout/index.jsx [hunk 2]
```diff
@@ -504,11 +504,21 @@ export default withCurrentRoute(
...
 
-		const colorScheme = getColorScheme( {
-			state,
-			sectionName,
-			isGlobalSidebarVisible,
-		} );
+		let colorScheme = null;
+
+		// Try to get the color scheme from `color_scheme` query param if we're in the WooCommerce Passwordless JPC flow.
+		if ( isWooPasswordlessJPC ) {
```
Comment (chihsuan) [inline client/layout/index.jsx]: How about using the ternary operator here to simplify the logic?

```js
let colorScheme =
    isWooPasswordlessJPC
        ? getColorSchemeFromCurrentQuery(currentQuery)
        : getColorScheme({
              state,
              isGlobalSidebarVisible,
              sectionName,
          });
```

File: client/layout/logged-out.jsx [hunk 3]
```diff
@@ -153,6 +156,10 @@ const LayoutLoggedOut = ( {
 
 	let masterbar = null;
 
+	useEffect( () => {
+		refreshColorScheme( 'default', colorScheme );
```
Comment (chihsuan) [inline client/layout/logged-out.jsx]: Should we check if `isWooPasswordlessJPC` before calling the function? If it's only needed for Woo.

File: client/layout/masterbar/woo.scss [hunk 4]
```diff
@@ -185,7 +185,7 @@ $breakpoint-mobile: 660px;
 
 		.progress-bar__progress {
 			border-radius: 0;
-			background-color: var(--wp-admin-theme-color);
+			background-color: var(--color-primary);
```
Comment (chihsuan) [inline client/layout/masterbar/woo.scss:188]: Nice fix!

**Review by chihsuan — COMMENTED** [review]
Tested well! 👍  I have some comments on the code.

File: client/layout/index.jsx (hunk 2 above)
Comment (moon0326) [reply client/layout/index.jsx]: I was tempted to use ternary at first.

When `color_scheme` query param  is missing for some reason, I think we should call `getColorScheme` to get the user's preference instead.

Comment (adrianduffell) [reply client/layout/index.jsx]: > I think we should call getColorScheme to get the user's preference instead.

That sounds good. I noticed there doesn't seem to be a global user preference on WPCOM anymore. We could perhaps also just default the standard WPCOM colors 🤔 

wpcom global color scheme

File: client/layout/masterbar/woo.scss [hunk 5]
```diff
@@ -2120,3 +2120,19 @@ $breakpoint-mobile: 660px;
 		}
 	}
 }
+
+// Color scheme 'blue' fix
+// The 'blue' theme color from wp-calypso does not match the 'blue' color from the local WordPress admin color scheme.
+// Fix the color scheme 'blue' to match the local WordPress admin color scheme on the new JPC page.
+body.is-blue {
```
Comment (adrianduffell) [inline client/layout/masterbar/woo.scss]: What color does wp-calypso render for the 'blue' scheme? I wonder if we might have it wrong on the WC core side -- I noticed when 'blue' is applied, WordPress renders buttons in a tan `#d9aa59` color, but in core profiler they are `#2a63a1`. 

Screen Shot 2024-11-20 at 10 48 25 pm

Screen Shot 2024-11-20 at 10 54 01 pm

**Review by adrianduffell — APPROVED** [review]
Thanks @moon0326, it is testing nice and smooth! Pre-approving once the comments are resolved 🚀

File: client/layout/masterbar/woo.scss (hunk 5 above)
Comment (moon0326) [reply client/layout/masterbar/woo.scss]: @adrianduffell Good catch. I think the core profiler has incorrect color. We've been using `--wp-admin-theme-color` for the button color, but it should be `--color-pirmary`. I'll double check and open a PR in WC repo.

File: client/layout/index.jsx (hunk 2 above)
Comment (moon0326) [reply client/layout/index.jsx]: @adrianduffell Good catch! Let's use the default color scheme in that case.

File: client/layout/logged-out.jsx (hunk 3 above)
Comment (moon0326) [reply client/layout/logged-out.jsx]: @chihsuan FIxed it in https://github.com/Automattic/wp-calypso/pull/96557/commits/c184c35b8762b85f6fe8761f61f17d557dfa0c52

File: client/layout/index.jsx (hunk 2 above)
Comment (moon0326) [reply client/layout/index.jsx]: @chihsuan @adrianduffell Fixed it in [Return 'default' color scheme](https://github.com/Automattic/wp-calypso/pull/96557/commits/ef3813fdd41011a18c8cfe2e7952ff4c46cf441a)

File: client/layout/masterbar/woo.scss (hunk 5 above)
Comment (moon0326) [reply client/layout/masterbar/woo.scss]: @adrianduffell 

> Good catch. I think the core profiler has incorrect color.

I take it back 😄 

Without this change, wp-calypso renders the exact same color as the core profiler. 
Screenshot 2024-11-20 at 12 56 33 PM

I've removed it in [Remove blue color fix change -- no longer needed](https://github.com/Automattic/wp-calypso/pull/96557/commits/89130bf5771b29435493ce540dc25d52e1b7dfae)

**Review by chihsuan — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**